tests/Microsoft.Identity.Test.Unit/AppTokenProviderTests.cs

```diff
@@ -0,0 +1,169 @@
+
+            var tokens = app.AppTokenCacheInternal.Accessor.GetAllAccessTokens();
+
+            Assert.AreEqual(1, tokens.Count);
+
+            var token = tokens.FirstOrDefault();
+            Assert.IsNotNull(token);
+            Assert.AreEqual(TestConstants.DefaultAccessToken, token.Secret);
+
+            // AcquireToken from cache
+            result = await app.AcquireTokenForClient(TestConstants.s_scope)
+                                                    .ExecuteAsync(new CancellationToken()).ConfigureAwait(false);
+
+            Assert.IsNotNull(result.AccessToken);
+            Assert.AreEqual(TestConstants.DefaultAccessToken, result.AccessToken);
+            Assert.AreEqual(TokenSource.Cache, result.AuthenticationResultMetadata.TokenSource);
+            Assert.AreEqual(1, callbackInvoked);
+
+            // Expire token
+            TokenCacheHelper.ExpireAllAccessTokens(app.AppTokenCacheInternal);
+
+            // Acquire token from app provider with expired token
+            result = await app.AcquireTokenForClient(TestConstants.s_scope)
+                                                    .ExecuteAsync(new CancellationToken()).ConfigureAwait(false);
+
+            Assert.IsNotNull(result.AccessToken);
+            Assert.AreEqual(TestConstants.DefaultAccessToken, result.AccessToken);
+            Assert.AreEqual(TokenSource.IdentityProvider, result.AuthenticationResultMetadata.TokenSource);
+            Assert.AreEqual(2, callbackInvoked);
+
+            differentScopesForAt = "new scope";
+
+            // Acquire token from app provider with new scopes
+            result = await app.AcquireTokenForClient(new[] { differentScopesForAt })
+                                                    .ExecuteAsync(new CancellationToken()).ConfigureAwait(false);
+
+            Assert.IsNotNull(result.AccessToken);
+            Assert.AreEqual(TestConstants.DefaultAccessToken + differentScopesForAt, result.AccessToken);
+            Assert.AreEqual(TokenSource.IdentityProvider, result.AuthenticationResultMetadata.TokenSource);
+            Assert.AreEqual(app.AppTokenCacheInternal.Accessor.GetAllAccessTokens().Count, 2);
+            Assert.AreEqual(3, callbackInvoked);
+
+            // Acquire token from app provider with claims. Should not use cache
+            result = await app.AcquireTokenForClient(TestConstants.s_scope)
+                                                    .WithClaims(TestConstants.Claims)
+                                                    .ExecuteAsync(new CancellationToken()).ConfigureAwait(false);
+
+            Assert.IsNotNull(result.AccessToken);
+            Assert.AreEqual(TestConstants.DefaultAccessToken + differentScopesForAt, result.AccessToken);
+            Assert.AreEqual(TokenSource.IdentityProvider, result.AuthenticationResultMetadata.TokenSource);
+            Assert.AreEqual(4, callbackInvoked);
+        }
+
+        [TestMethod]
+        public async Task CheckRefreshInAsync()
```

Comment:
- Missing test case when AppTokenProviderResult.RefreshInSeconds is set.
- Can this be refactored to use DataRows like [here](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/blob/8a0a64008bb6053cde73a48bf8f8c7ae8faf4664/tests/Microsoft.Identity.Test.Unit/ManagedIdentityTests/ManagedIdentityTests.cs#L447-L452)? Maybe pass in expiration time and expected RefreshOn.HasValue